packages/typespec-ts/src/modular/emitModels.ts

```diff
@@ -517,36 +519,50 @@ function addExtendedDictInfo(
   const additionalPropertiesType = model.additionalProperties
     ? getTypeExpression(context, model.additionalProperties)
     : undefined;
-  if (
-    (model.properties &&
-      model.properties.length > 0 &&
-      model.additionalProperties &&
-      model.properties?.every((p) => {
+  if (context.rlcOptions?.compatibilityMode) {
+    const ancestors = getAllAncestors(model);
+    const properties = getAllProperties(model, ancestors);
+    let anyType = true;
+    if (!additionalPropertiesType) {
+      // case 1: if additionalProperties is not defined, we should use any type
+      anyType = true;
+    } else if (properties.length === 0) {
+      // case 2: if additionalProperties is defined and model.properties is empty, we should use additionalProperties type
+      anyType = false;
+    } else {
+      // case 3: if additionalProperties is defined and model.properties is not empty, we should check if all properties are compatible with additionalProperties type
+      anyType = !properties.every((p) => {
```

Comment:
I think this would return false for properties.length === 0's case?

Reply:
Do you mean for case line 529 - 531? yes if there is no any model property we would directly trust the type defined in additional record.